deepchem/models/tensorgraph/tensor_graph.py

```diff
@@ -238,123 +229,146 @@ def default_generator(self,
-      processing the samples in a batched way.
-
-    # Arguments
-        x: the input data, as a Numpy array.
-        batch_size: integer.
-        verbose: verbosity mode, 0 or 1.
-
-    # Returns
-        A Numpy array of predictions.
-    """
-    retval = self.predict_proba_on_generator(generator, transformers)
-    if self.mode == 'classification':
-      retval = np.expand_dims(from_one_hot(retval, axis=2), axis=1)
-    return retval
-
-  def predict_proba_on_generator(self, generator, transformers=[]):
+  def predict_on_generator(self, generator, transformers=[], outputs=None):
     """
+    Parameters
+    ----------
+    generator: Generator
+      Generator that constructs feed dictionaries for TensorGraph.
+    transformers: list
+      List of dc.trans.Transformers.
+    outputs: object
+      If outputs is None, then will assume outputs = self.outputs.
+      If outputs is a Layer/Tensor, then will evaluate and return as a
+      single ndarray. If outputs is a list of Layers/Tensors, will return a list
+      of ndarrays.
     Returns:
       y_pred: numpy ndarray of shape (n_samples, n_classes*n_tasks)
     """
     if not self.built:
       self.build()
+    if outputs is None:
+      outputs = self.outputs
+    elif not isinstance(outputs, list):
+      outputs = [outputs]
     with self._get_tf("Graph").as_default():
       with tf.Session() as sess:
         saver = tf.train.Saver()
         self._initialize_weights(sess, saver)
         out_tensors = [x.out_tensor for x in self.outputs]
-        results = []
+        # Gather results for each output
+        results = [[] for out in out_tensors]
         for feed_dict in generator:
           feed_dict = {
               self.layers[k.name].out_tensor: v
               for k, v in six.iteritems(feed_dict)
           }
           feed_dict[self._training_placeholder] = 0.0
-          result = np.array(sess.run(out_tensors, feed_dict=feed_dict))
-          if len(result.shape) == 3:
-            result = np.transpose(result, axes=[1, 0, 2])
-          result = undo_transforms(result, transformers)
-          results.append(result)
-        return np.concatenate(results, axis=0)
-
-  def bayesian_predict_on_batch(self, X, transformers=[], n_passes=4):
+          feed_results = sess.run(out_tensors, feed_dict=feed_dict)
+          if len(feed_results) > 1:
+            if len(transformers):
+              raise ValueError("Does not support transformations "
+                               "for multiple outputs.")
+          elif len(feed_results) == 1:
+            result = undo_transforms(feed_results[0], transformers)
+            feed_results = [result]
+          for ind, result in enumerate(feed_results):
+            results[ind].append(result)
+
+        final_results = []
+        for result_list in results:
+          final_results.append(np.concatenate(result_list, axis=0))
+        # If only one output, just return array
+        if len(final_results) == 1:
```

Comment:
This decision should be based on what they specified for `outputs`, not based on the number of outputs.  If they specify `outputs=myoutput`, that should return an array, but if they specify `outputs=[myoutput]`. that should return a list with one element.